apollo-runtime/src/main/java/com/apollographql/android/cache/normalized/lru/EvictionPolicy.java

```diff
@@ -0,0 +1,117 @@
+
+import com.apollographql.android.api.graphql.internal.Optional;
+
+import java.util.concurrent.TimeUnit;
+
+import static android.content.Context.ACTIVITY_SERVICE;
+import static android.content.pm.ApplicationInfo.FLAG_LARGE_HEAP;
+
+public final class EvictionPolicy {
+
+  private final Optional<Long> maxSizeBytes;
+  private final Optional<Long> maxEntries;
+  private final Optional<Long> expireAfterAccess;
+  private final Optional<TimeUnit> expireAfterAccessTimeUnit;
+  private final Optional<Long> expireAfterWrite;
+  private final Optional<TimeUnit> expireAfterWriteTimeUnit;
+
+  public Optional<Long> maxSizeBytes() {
+    return maxSizeBytes;
+  }
+
+  public Optional<Long> maxEntries() {
+    return maxEntries;
+  }
+
+  public Optional<Long> expireAfterAccess() {
+    return expireAfterAccess;
+  }
+
+  public Optional<TimeUnit> expireAfterAccessTimeUnit() {
+    return expireAfterAccessTimeUnit;
+  }
+
+  public Optional<Long> expireAfterWrite() {
+    return expireAfterWrite;
+  }
+
+  public Optional<TimeUnit> expireAfterWriteTimeUnit() {
+    return expireAfterWriteTimeUnit;
+  }
+
+  public static EvictionPolicy.Builder builder() {
+    return new EvictionPolicy.Builder();
+  }
+
+  public static class Builder {
+
+    private Builder() { }
+
+    private Optional<Long> maxSizeBytes = Optional.absent();
+    private Optional<Long> maxEntries = Optional.absent();
+    private Optional<Long> expireAfterAccess = Optional.absent();
+    private Optional<TimeUnit> expireAfterAccessTimeUnit = Optional.absent();
+    private Optional<Long> expireAfterWrite = Optional.absent();
+    private Optional<TimeUnit> expireAfterWriteTimeUnit = Optional.absent();
+
+    public EvictionPolicy.Builder deviceBasedMaxSize(Context context) {
+      this.maxSizeBytes = Optional.of(calculateDeviceMemoryCacheSize(context));
+      return this;
+    }
+
+    public EvictionPolicy.Builder maxSizeBytes(long maxSizeBytes) {
+      this.maxSizeBytes = Optional.of(maxSizeBytes);
+      return this;
+    }
+
+    public EvictionPolicy.Builder maxEntries(long maxEntries) {
+      this.maxEntries = Optional.of(maxEntries);
+      return this;
+    }
+
+    public EvictionPolicy.Builder expireAfterAccess(long time, TimeUnit timeUnit) {
+      this.expireAfterAccess = Optional.of(time);
+      this.expireAfterAccessTimeUnit = Optional.of(timeUnit);
+      return this;
+    }
+
+    public EvictionPolicy.Builder expireAfterWrite(long time, TimeUnit timeUnit) {
+      this.expireAfterWrite = Optional.of(time);
+      this.expireAfterWriteTimeUnit = Optional.of(timeUnit);
+      return this;
+    }
+
+    public EvictionPolicy build() {
+      return new EvictionPolicy(maxSizeBytes, maxEntries, expireAfterAccess, expireAfterAccessTimeUnit,
+          expireAfterWrite, expireAfterWriteTimeUnit);
+    }
+
+  }
+
+  /**
+   * Derived from https://github.com/square/picasso cache sizing.
+   */
+  private static long calculateDeviceMemoryCacheSize(Context context) {
```

Comment:
Not entirely sure but it seems that those values are just dalvik VM [properties](https://android.googlesource.com/platform/frameworks/base/+/742a67127366c376fdf188ff99ba30b27d3bf90c/core/java/android/app/ActivityManager.java#181). Are those accessible through a `System.getProperty()`?

The closest thing we can get to without context is `Runtime.getRuntime().maxMemory` which gives the number of bytes in heap instead of the recommended usage so maybe use that and reduce the target even more?